Display warning/error when building android application with build number greater than allowed value

Hello everyone. I have setup a CI which uses the current timestamp since epoch as the build number. It was working just fine, until today. It stopped accepting the build-number. I suppose this is a problem at handling large integer values ? Btw, our app is actually blocked upon compiling (and we cannot go back as we already published several playstore and app-store versions).

## Steps to Reproduce

1. Run `flutter create bug`.
2. flutter build appbundle -t lib/main.dart --build-name=5.44.0 --build-number=1590170586079 --release --tree-shake-icons

**Expected results:** AppBundle or APK compiles

**Actual results:** 
```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/nicolas/dev/xxx/android/app/build.gradle' line: 42

* What went wrong:
A problem occurred evaluating project ':app'.
> For input string: "1590170586079"

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 663ms
Running Gradle task 'bundleXXX'...
Running Gradle task 'bundleXXX'... Done                  1,4s
Gradle task bundleXXX failed with exit code 1
```

and line 42 is: 
```
versionCode flutterVersionCode.toInteger()
```

<details>
  <summary>Logs</summary>

<!-- Finally, paste the output of running `flutter doctor -v` here. -->

```
flutter doctor -v
[✓] Flutter (Channel stable, v1.17.1, on Mac OS X 10.15.4 19E287, locale it-IT)
    • Flutter version 1.17.1 at /Users/nicolas/dev/flutter_sdk/flutter
    • Framework revision f7a6a7906b (10 days ago), 2020-05-12 18:39:00 -0700
    • Engine revision 6bc433c6b6
    • Dart version 2.8.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nicolas/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Users/nicolas/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/192.6392135/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Android Studio (version 3.6)
    • Android Studio at /Users/nicolas/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/192.6392135/Android
      Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[!] IntelliJ IDEA Ultimate Edition (version 2020.1.1)
    • IntelliJ at /Users/nicolas/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    • Dart plugin version 201.7223.99
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.42.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.0

[✓] Connected device (1 available)
    • ASUS Z00A • XXX • android-x86 • Android 7.1.2 (API 25)

! Doctor found issues in 1 category.
```

</details>

--- a/packages/flutter_tools/lib/src/android/build_validation.dart
+++ b/packages/flutter_tools/lib/src/android/build_validation.dart
@@ -1,0 +1,38 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../base/common.dart';
+import '../build_info.dart';
+
+const String kGooglePlayVersioning = 'https://developer.android.com/studio/publish/versioning.html';
+
+/// Validates that the build mode and build number are valid for a given build.
+void validateBuild(AndroidBuildInfo androidBuildInfo) {
+  final BuildInfo buildInfo = androidBuildInfo.buildInfo;
+  if (buildInfo.mode.isPrecompiled && androidBuildInfo.targetArchs.contains(AndroidArch.x86)) {
+    throwToolExit('Cannot build ${androidBuildInfo.buildInfo.mode.name} mode for x86 ABI.');
+  }
+  if (buildInfo.buildNumber != null) {
+    final int result = int.tryParse(buildInfo.buildNumber);
+    if (result == null) {
+      throwToolExit(
+        'buildNumber: ${buildInfo.buildNumber} was not a valid integer value.\n'
+        'For more information see $kGooglePlayVersioning .'
+      );
+    }
+    if (result < 0) {
+      throwToolExit(
+        'buildNumber: ${buildInfo.buildNumber} must be a positive integer value.\n'
+        'For more information see $kGooglePlayVersioning .'
+      );
+    }
+    if (result > 2100000000) {
+      throwToolExit(
+        'buildNumber: ${buildInfo.buildNumber} is greater than the maximum '
+        'allowed value of 2100000000.\n'
+        'For more information see $kGooglePlayVersioning .'
+      );
+    }
+  }
+}

--- a/packages/flutter_tools/lib/src/android/build_validation.dart
+++ b/packages/flutter_tools/lib/src/android/build_validation.dart
@@ -6,12 +6,16 @@ import '../base/common.dart';
 import '../build_info.dart';
 
 const String kGooglePlayVersioning = 'https://developer.android.com/studio/publish/versioning.html';
+const String kSupportedAbis = 'https://flutter.dev/docs/deployment/android#what-are-the-supported-target-architectures';
 
 /// Validates that the build mode and build number are valid for a given build.
 void validateBuild(AndroidBuildInfo androidBuildInfo) {
   final BuildInfo buildInfo = androidBuildInfo.buildInfo;
   if (buildInfo.mode.isPrecompiled && androidBuildInfo.targetArchs.contains(AndroidArch.x86)) {
-    throwToolExit('Cannot build ${androidBuildInfo.buildInfo.mode.name} mode for x86 ABI.');
+    throwToolExit(
+      'Cannot build ${androidBuildInfo.buildInfo.mode.name} mode for x86 ABI.\n'
+      'For more information see $kSupportedAbis .'
+    );
   }
   if (buildInfo.buildNumber != null) {
     final int result = int.tryParse(buildInfo.buildNumber);

--- a/packages/flutter_tools/lib/src/commands/build_apk.dart
+++ b/packages/flutter_tools/lib/src/commands/build_apk.dart
@@ -5,6 +5,7 @@
 import 'dart:async';
 
 import '../android/android_builder.dart';
+import '../android/build_validation.dart';
 import '../android/gradle_utils.dart';
 import '../base/terminal.dart';
 import '../build_info.dart';
@@ -95,6 +96,7 @@ class BuildApkCommand extends BuildSubCommand {
       targetArchs: stringsArg('target-platform').map<AndroidArch>(getAndroidArchForName),
       shrink: boolArg('shrink'),
     );
+    validateBuild(androidBuildInfo);
 
     if (buildInfo.isRelease && !androidBuildInfo.splitPerAbi && androidBuildInfo.targetArchs.length > 1) {
       final String targetPlatforms = stringsArg('target-platform').join(', ');

--- a/packages/flutter_tools/lib/src/commands/build_appbundle.dart
+++ b/packages/flutter_tools/lib/src/commands/build_appbundle.dart
@@ -5,6 +5,7 @@
 import 'dart:async';
 
 import '../android/android_builder.dart';
+import '../android/build_validation.dart';
 import '../android/gradle_utils.dart';
 import '../build_info.dart';
 import '../cache.dart';
@@ -83,6 +84,7 @@ class BuildAppBundleCommand extends BuildSubCommand {
       targetArchs: stringsArg('target-platform').map<AndroidArch>(getAndroidArchForName),
       shrink: boolArg('shrink'),
     );
+    validateBuild(androidBuildInfo);
     await androidBuilder.buildAab(
       project: FlutterProject.current(),
       target: targetFile,

--- a/packages/flutter_tools/test/general.shard/android/build_validation_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/build_validation_test.dart
@@ -1,0 +1,64 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter_tools/src/android/build_validation.dart';
+import 'package:flutter_tools/src/build_info.dart';
+
+import '../../src/common.dart';
+
+void main() {
+  testWithoutContext('validateBuild throws if attempting to build release/profile on x86', () {
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        BuildInfo.release,
+        targetArchs: <AndroidArch>[AndroidArch.x86],
+      ),
+    ), throwsToolExit(message: 'Cannot build release mode for x86 ABI.'));
+  });
+
+  testWithoutContext('validateBuild does not throw on AOT supported architectures', () {
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        BuildInfo.release,
+        targetArchs: <AndroidArch>[AndroidArch.x86_64, AndroidArch.armeabi_v7a, AndroidArch.arm64_v8a],
+      ),
+    ), returnsNormally);
+  });
+
+  testWithoutContext('validateBuild throws if an invalid build number is specified', () {
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        // Invalid number
+        BuildInfo(BuildMode.debug, '', treeShakeIcons: false, buildNumber: 'a'),
+        targetArchs: <AndroidArch>[AndroidArch.x86],
+      ),
+    ), throwsToolExit(message: 'buildNumber: a was not a valid integer value.'));
+
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        // Negative number
+        BuildInfo(BuildMode.debug, '', treeShakeIcons: false, buildNumber: '-1'),
+        targetArchs: <AndroidArch>[AndroidArch.x86],
+      ),
+    ), throwsToolExit(message: 'buildNumber: -1 must be a positive integer value.'));
+
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        // Nigger than maximum supported play store value
+        BuildInfo(BuildMode.debug, '', treeShakeIcons: false, buildNumber: '2100000001'),
+        targetArchs: <AndroidArch>[AndroidArch.x86],
+      ),
+    ), throwsToolExit(message: 'buildNumber: 2100000001 is greater than the maximum '
+        'allowed value of 2100000000.'));
+  });
+
+  testWithoutContext('validateBuild does not throw on positive number', () {
+    expect(() => validateBuild(
+      const AndroidBuildInfo(
+        BuildInfo(BuildMode.debug, '', treeShakeIcons: false, buildNumber: '2'),
+        targetArchs: <AndroidArch>[AndroidArch.x86],
+      ),
+    ), returnsNormally);
+  });
+}

